File: validation/python/generate_gold_standard.py
#!/usr/bin/env python3
"""Independent numerical validation for the browser reliability engine.

The application remains a custom TypeScript implementation. This script uses
NumPy and SciPy's independently maintained probability distributions, then
checks every reported metric (including ICC confidence limits) against both
R reference output and TypeScript output. It intentionally does not claim that
Pingouin was executed.
"""

from __future__ import annotations

import argparse
import csv
import json
import math
import platform
from datetime import datetime, timezone
from pathlib import Path

import numpy as np
import scipy
from scipy import stats


ROOT = Path(__file__).resolve().parents[1]
DATASETS_DIR = ROOT / "datasets"
FIELDS = (
    "n",
    "t1_mean",
    "t2_mean",
    "grand_mean",
    "mean_bias",
    "sd_diff",
    "typical_error",
    "cv_percent",
    "pooled_sd",
    "icc_a1",
    "icc_a1_lower95",
    "icc_a1_upper95",
    "sem",
    "mdc95",
    "mdc_percent",
    "loa_lower",
    "loa_upper",
)

# R fixtures are rounded to four decimals; this tolerance is deliberately
# tighter than any product decision threshold but allows fixture rounding.
R_ABSOLUTE_TOLERANCE = 5e-4
TS_ABSOLUTE_TOLERANCE = 5e-5


def read_dataset(csv_path: Path) -> tuple[np.ndarray, np.ndarray]:
    with csv_path.open("r", encoding="utf-8", newline="") as handle:
        rows = list(csv.DictReader(handle))
    return (
        np.asarray([float(row["test1"]) for row in rows], dtype=float),
        np.asarray([float(row["test2"]) for row in rows], dtype=float),
    )


def calculate_results(t1: np.ndarray, t2: np.ndarray) -> dict[str, object]:
    if t1.shape != t2.shape or t1.size < 2:
        raise ValueError("Validation data require at least two paired observations.")
    if not np.isfinite(t1).all() or not np.isfinite(t2).all():
        raise ValueError("Validation data must contain only finite values.")

    n = int(t1.size)
    k = 2
    matrix = np.column_stack((t1, t2))
    t1_mean = float(t1.mean())
    t2_mean = float(t2.mean())
    grand_mean = float(matrix.mean())
    row_means = matrix.mean(axis=1)
    column_means = matrix.mean(axis=0)

    ss_total = float(np.square(matrix - grand_mean).sum())
    ss_rows = float(k * np.square(row_means - grand_mean).sum())
    ss_columns = float(n * np.square(column_means - grand_mean).sum())
    ss_error = max(0.0, ss_total - ss_rows - ss_columns)
    ms_rows = ss_rows / (n - 1)
    ms_columns = ss_columns / (k - 1)
    ms_error = ss_error / ((n - 1) * (k - 1))

    icc_denominator = (
        ms_rows
        + (k - 1) * ms_error
        + (k / n) * (ms_columns - ms_error)
    )
    icc_a1 = (ms_rows - ms_error) / icc_denominator

    # McGraw & Wong ICC(A,1) exact limits. SciPy supplies the F quantile;
    # R irr::icc is the independent package-level reference.
    f_upper = float(stats.f.ppf(0.975, n - 1, n - 1))
    lower_numerator = n * (ms_rows - f_upper * ms_error)
    lower_denominator = (
        f_upper * (k * ms_columns + (k * n - k - n) * ms_error)
        + n * ms_rows
    )
    upper_numerator = n * (f_upper * ms_rows - ms_error)
    upper_denominator = (
        k * ms_columns
        + (k * n - k - n) * ms_error
        + n * f_upper * ms_rows
    )
    lower_ci = float(np.clip(lower_numerator / lower_denominator, -1.0, 1.0))
    upper_ci = float(np.clip(upper_numerator / upper_denominator, -1.0, 1.0))

    differences = t2 - t1
    mean_bias = float(differences.mean())
    sd_diff = float(differences.std(ddof=1))
    typical_error = sd_diff / math.sqrt(2)
    pooled_sd = float(
        math.sqrt(
            ((n - 1) * t1.var(ddof=1) + (n - 1) * t2.var(ddof=1))
            / (2 * n - 2)
        )
    )
    sem = pooled_sd * math.sqrt(max(0.0, 1.0 - icc_a1))
    mdc95 = sem * float(stats.norm.ppf(0.975)) * math.sqrt(2)
    cv_percent = typical_error / abs(grand_mean) * 100 if abs(grand_mean) >= 1e-6 else None
    mdc_percent = mdc95 / abs(grand_mean) * 100 if abs(grand_mean) >= 1e-6 else None

    return {
        "n": n,
        "t1_mean": t1_mean,
        "t2_mean": t2_mean,
        "grand_mean": grand_mean,
        "mean_bias": mean_bias,
        "sd_diff": sd_diff,
        "typical_error": typical_error,
        "cv_percent": cv_percent,
        "pooled_sd": pooled_sd,
        "icc_a1": float(icc_a1),
        "icc_a1_lower95": lower_ci,
        "icc_a1_upper95": upper_ci,
        "sem": sem,
        "mdc95": mdc95,
        "mdc_percent": mdc_percent,
        "loa_lower": mean_bias - 1.959964 * sd_diff,
        "loa_upper": mean_bias + 1.959964 * sd_diff,
        "anova_table": {
            "ms_rows": ms_rows,
            "ms_columns": ms_columns,
            "ms_error": ms_error,
            "df_rows": n - 1,
            "df_columns": k - 1,
            "df_error": (n - 1) * (k - 1),
        },
    }


def payload(dataset_id: str, results: dict[str, object]) -> dict[str, object]:
    return {
        "dataset_id": dataset_id,
        "engine": "Python-NumPy-SciPy-Independent-Validation",
        "python_version": platform.python_version(),
        "packages": {"numpy": np.__version__, "scipy": scipy.__version__},
        "model_specification": {
            "design": "Two-way random-effects ANOVA",
            "definition": "Absolute agreement",
            "measure": "Single score ICC(A,1) / ICC(2,1)",
            "ci_method": "McGraw & Wong exact F-distribution; scipy.stats.f.ppf",
            "independent_primary_reference": "R irr::icc(model='twoway', type='agreement', unit='single')",
        },
        "results": results,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }


def assert_close(
    dataset_id: str,
    source: str,
    actual: dict[str, object],
    expected: dict[str, object],
    tolerance: float,
) -> None:
    failures: list[str] = []
    for field in FIELDS:
        observed = actual.get(field)
        reference = expected.get(field)
        if observed is None and reference is None:
            continue
        if observed is None or reference is None:
            failures.append(f"{field}: {observed!r} != {reference!r}")
            continue
        difference = abs(float(observed) - float(reference))
        if difference > tolerance:
            failures.append(
                f"{field}: observed={float(observed):.9g}, "
                f"reference={float(reference):.9g}, diff={difference:.3g}"
            )
    if failures:
        details = "\n  ".join(failures)
        raise AssertionError(f"{dataset_id} failed against {source}:\n  {details}")


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--check",
        action="store_true",
        help="Do not rewrite fixtures; fail if Python differs from R or TypeScript.",
    )
    args = parser.parse_args()

    checked = 0
    for dataset_dir in sorted(path for path in DATASETS_DIR.iterdir() if path.is_dir()):
        t1, t2 = read_dataset(dataset_dir / "input.csv")
        results = calculate_results(t1, t2)
        output = payload(dataset_dir.name, results)

        if args.check:
            r_results = json.loads((dataset_dir / "expected_R.json").read_text(encoding="utf-8"))["results"]
            ts_results = json.loads((dataset_dir / "actual_typescript.json").read_text(encoding="utf-8"))["results"]
            assert_close(dataset_dir.name, "R irr::icc fixtures", results, r_results, R_ABSOLUTE_TOLERANCE)
            assert_close(dataset_dir.name, "TypeScript engine", results, ts_results, TS_ABSOLUTE_TOLERANCE)
        else:
            (dataset_dir / "expected_Python.json").write_text(
                json.dumps(output, indent=2, ensure_ascii=False) + "\n",
                encoding="utf-8",
            )
        checked += 1

    action = "validated" if args.check else "generated"
    print(f"Python/SciPy {action} {checked} datasets, including ICC 95% confidence limits.")


if __name__ == "__main__":
    main()

File: validation/R/generate_gold_standard.R
# ==============================================================================
# Gold Standard Test-Retest Reliability Independent R Validation Script
# ==============================================================================
# Methodology Reference:
# 1. McGraw, K. O., & Wong, S. P. (1996). Forming inferences about some intraclass
#    correlation coefficients. Psychological Methods, 1(1), 30–46.
# 2. Weir, J. P. (2005). Quantifying test-retest reliability using the intraclass
#    correlation coefficient and the SEM. J Strength Cond Res, 19(1), 231-240.
# 3. Hopkins, W. G. (2000). Measures of reliability in sports medicine and science.
#    Sports Med, 30(1), 1-15.
# ==============================================================================

suppressPackageStartupMessages({
  if (!require("irr", quietly = TRUE)) install.packages("irr", repos = "https://cloud.r-project.org")
  if (!require("psych", quietly = TRUE)) install.packages("psych", repos = "https://cloud.r-project.org")
  if (!require("jsonlite", quietly = TRUE)) install.packages("jsonlite", repos = "https://cloud.r-project.org")
  library(irr)
  library(psych)
  library(jsonlite)
})

# Package versions metadata
r_version <- paste(R.version$major, R.version$minor, sep = ".")
irr_version <- as.character(packageVersion("irr"))
psych_version <- as.character(packageVersion("psych"))

dataset_dirs <- list.dirs("validation/datasets", full.names = TRUE, recursive = FALSE)

cat(sprintf("=== Running Independent R Validation Suite ===\n"))
cat(sprintf("R Version: %s\n", r_version))
cat(sprintf("irr Package Version: %s\n", irr_version))
cat(sprintf("psych Package Version: %s\n\n", psych_version))

for (dir in dataset_dirs) {
  csv_file <- file.path(dir, "input.csv")
  if (!file.exists(csv_file)) next
  
  df <- read.csv(csv_file, stringsAsFactors = FALSE)
  dataset_id <- basename(dir)
  
  # Standardize column structure
  t1 <- df$test1
  t2 <- df$test2
  ratings_mat <- cbind(t1, t2)
  n <- nrow(df)
  
  # --------------------------------------------------------------------------
  # 1. ICC Model Execution via irr::icc
  # Model: "twoway" (Two-way random-effects model for absolute agreement)
  # Type: "agreement" (Absolute agreement definition: includes systematic column variance in denominator)
  # Unit: "single" (Single measure reliability ICC(A,1))
  # --------------------------------------------------------------------------
  icc_irr <- irr::icc(
    ratings_mat,
    model = "twoway",
    type = "agreement",
    unit = "single",
    conf.level = 0.95
  )
  
  # --------------------------------------------------------------------------
  # 2. ICC Verification via psych::ICC
  # In psych::ICC output table, row 'ICC2' corresponds to:
  # "Two-way random effects, single rater, absolute agreement" [ICC(A,1) / ICC(2,1)].
  # Note: psych::ICC labels row ICC3 as "Two-way mixed, consistency", NOT absolute agreement.
  # --------------------------------------------------------------------------
  icc_psych <- psych::ICC(ratings_mat)
  icc2_row <- icc_psych$results[icc_psych$results$type == "ICC2", ]
  
  # --------------------------------------------------------------------------
  # 3. Descriptive & Hopkins / Weir Metrics
  # --------------------------------------------------------------------------
  t1_mean <- mean(t1)
  t2_mean <- mean(t2)
  grand_mean <- mean(c(t1, t2))
  diffs <- t2 - t1
  mean_bias <- mean(diffs)
  sd_diff <- sd(diffs)
  
  # Hopkins Typical Error
  te <- sd_diff / sqrt(2)
  
  # Coefficient of Variation (%)
  cv_mean <- (te / grand_mean) * 100
  
  # Weir (2005) SEM = pooled_sd * sqrt(1 - ICC_A1)
  # Pooled SD across both trials
  ss1 <- sum((t1 - t1_mean)^2)
  ss2 <- sum((t2 - t2_mean)^2)
  pooled_sd <- sqrt((ss1 + ss2) / (2 * n - 2))
  
  icc_point <- icc_irr$value
  sem <- pooled_sd * sqrt(max(0, 1 - icc_point))
  
  # Minimal Detectable Change (MDC95)
  mdc95 <- sem * 1.96 * sqrt(2)
  mdc_percent <- (mdc95 / grand_mean) * 100
  
  # Bland-Altman 95% Limits of Agreement
  loa_lower <- mean_bias - 1.96 * sd_diff
  loa_upper <- mean_bias + 1.96 * sd_diff
  
  output <- list(
    dataset_id = dataset_id,
    engine = "R-Statistical-Core",
    r_version = r_version,
    packages = list(irr = irr_version, psych = psych_version),
    icc_model_spec = list(
      irr_call = "irr::icc(ratings_mat, model='twoway', type='agreement', unit='single', conf.level=0.95)",
      psych_row = "ICC2 (Two-way random single-rater absolute agreement)",
      mcgraw_wong_notation = "ICC(A,1)",
      shrout_fleiss_notation = "ICC(2,1)"
    ),
    results = list(
      n = n,
      t1_mean = round(t1_mean, 6),
      t2_mean = round(t2_mean, 6),
      grand_mean = round(grand_mean, 6),
      mean_bias = round(mean_bias, 6),
      sd_diff = round(sd_diff, 6),
      typical_error = round(te, 6),
      cv_percent = round(cv_mean, 6),
      pooled_sd = round(pooled_sd, 6),
      icc_a1 = round(icc_point, 6),
      icc_a1_lower95 = round(icc_irr$lbound, 6),
      icc_a1_upper95 = round(icc_irr$ubound, 6),
      sem = round(sem, 6),
      mdc95 = round(mdc95, 6),
      mdc_percent = round(mdc_percent, 6),
      loa_lower = round(loa_lower, 6),
      loa_upper = round(loa_upper, 6)
    ),
    timestamp = Sys.time()
  )
  
  out_path <- file.path(dir, "expected_R.json")
  write_json(output, out_path, pretty = TRUE, auto_unbox = TRUE)
  cat(sprintf("[R Validation] Generated %s -> ICC(A,1) = %.6f [%.6f, %.6f]\n", dataset_id, icc_point, icc_irr$lbound, icc_irr$ubound))
}

cat("\n=== R Validation Completed Successfully ===\n")
